src/ava-fast-check.ts

```diff
@@ -52,40 +90,30 @@ function internalTestProp<Ts extends [any] | any[]>(
   testFn(`${label} (with seed=${customParams.seed})`, wrapProp(arbitraries, prop, params));
 }
 
-export function testProp<Ts extends [any] | any[]>(
-  label: string,
-  arbitraries: ArbitraryTuple<Ts>,
-  prop: Prop<Ts>,
-  params?: fc.Parameters<Ts>
-): void {
-  internalTestProp(test, label, arbitraries, prop, params);
+function exposeModifier<Context, T extends Extract<keyof TestInterface, AvaModifierWhitelist>>(
+  modifier: T
+): PropertyTest<Context> {
+  return (label, arbitraries, prop, params) =>
+    internalTestProp((test as TestInterface<Context>)[modifier], label, arbitraries, prop, params);
 }
 
-export namespace testProp {
-  export const only = <Ts extends [any] | any[]>(
+export const testProp: PropertyTestInterface<unknown> = Object.assign(
+  function testProp<Context, Ts extends NonEmptyArray<any>>(
     label: string,
     arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
+    prop: Prop<Context, Ts>,
     params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.only, label, arbitraries, prop, params);
-  export const failing = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.failing, label, arbitraries, prop, params);
-  export const skip = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.skip, label, arbitraries, prop, params);
-  export const serial = <Ts extends [any] | any[]>(
-    label: string,
-    arbitraries: ArbitraryTuple<Ts>,
-    prop: Prop<Ts>,
-    params?: fc.Parameters<Ts>
-  ): void => internalTestProp(test.serial, label, arbitraries, prop, params);
-}
-
-export { test, fc };
+  ): void {
+    internalTestProp(test as TestInterface<Context>, label, arbitraries, prop, params);
+  },
+  {
+    only: exposeModifier('only'),
+    failing: exposeModifier('failing'),
+    skip: exposeModifier('skip'),
+    serial: exposeModifier('serial'),
+    before: test.before,
+    beforeEach: test.beforeEach,

```

Comment:
Ideally, test.beforeEach should be plugged to beforeEach of the property.

Indeed, according to the [documentation of AVA](https://github.com/avajs/ava/blob/master/docs/01-writing-tests.md#before--after-hooks): `test.beforeEach` is supposed to be executed before each test. A single run of a property might be considered as a test IMO.

Not merging the beforeEach of fast-check (on property) with the one of AVA might be misleading for the user.

```js
testProp.beforeEach(() => console.log('ava.beforeEach'))
testProp(
  'label',
  [fc.nat()],
  () => { console.log('one test');
  { beforeEach() { console.log('fc.beforeEach'); }}
);
// ava.beforeEach
// fc.beforeEach
// one test
// fc.beforeEach
// one test
// fc.beforeEach
// one test
// ...
```

For me we should also call the beforeEach of ava 🤔 Otherwise the spec of the beforeEach of ava does not seem to be fulfilled.

Reply:
Or mayby we can just not expose `beforeEach` for the moment to avoid such confusion...

Reply:
Agreed! I have removed the `beforeEach` and `afterEach` hooks for now since I believe you are right: each run of the property should be considered a single test, and offering a different hook is just confusing. 

I like the syntax you've proposed here, I created a new issue to reference this feature: https://github.com/dubzzz/ava-fast-check/issues/19